tests/e2e/e2e_setup_test.go

```diff
@@ -391,84 +387,116 @@ func noRestart(config *docker.HostConfig) {
 	}
 }
 
-func (s *IntegrationTestSuite) initUpgrade() {
+func (s *IntegrationTestSuite) upgrade() {
 	// submit, deposit, and vote for upgrade proposal
-	s.submitProposal(s.chains[0])
-	s.submitProposal(s.chains[1])
+	currentHeightA := s.chainHeight(s.valResources[s.chains[0].ChainMeta.Id][0].Container.ID)
+	currentHeightB := s.chainHeight(s.valResources[s.chains[1].ChainMeta.Id][0].Container.ID)
+	propHeightA = currentHeightA + 40 + len(s.chains[0].Validators)
+	propHeightB = currentHeightB + 40 + len(s.chains[1].Validators)
+
+	s.submitProposal(s.chains[0], propHeightA)
+	s.submitProposal(s.chains[1], propHeightB)
 	s.depositProposal(s.chains[0])
 	s.depositProposal(s.chains[1])
-	s.voteProposal(s.chains[0])
-	s.voteProposal(s.chains[1])
-
+	var wg sync.WaitGroup
+	wg.Add(2)
+	// use goroutines to vote on both chains concurrently (no possibility of sequence mismatch)
+	go s.voteProposal(s.chains[0], &wg)
+	go s.voteProposal(s.chains[1], &wg)
+	wg.Wait()
 	// wait till all chains halt at upgrade height
-	for _, chain := range s.chains {
-		for i := range chain.Validators {
-			s.T().Logf("waiting to reach upgrade height on %s validator container: %s", chain.ChainMeta.Id, s.valResources[chain.ChainMeta.Id][i].Container.ID)
+	for _, c := range s.chains {
+		if c.ChainMeta.Id == chain.ChainAID {
+			propHeight = propHeightA
+		} else {
+			propHeight = propHeightB
+		}
+		for i := range c.Validators {
+			// use counter to ensure no new blocks are being created
+			counter := 0
+			s.T().Logf("waiting to reach upgrade height on %s validator container: %s", s.valResources[c.ChainMeta.Id][i].Container.Name[1:], s.valResources[c.ChainMeta.Id][i].Container.ID)
 			s.Require().Eventually(
 				func() bool {
-					out := s.chainStatus(s.valResources[chain.ChainMeta.Id][i].Container.ID)
-					var syncInfo syncInfo
-					json.Unmarshal(out, &syncInfo)
-					if syncInfo.SyncInfo.LatestHeight != "75" {
-						s.T().Logf("current block height is %v, waiting for block 75 container: %s", syncInfo.SyncInfo.LatestHeight, s.valResources[chain.ChainMeta.Id][i].Container.ID)
+					currentHeight := s.chainHeight(s.valResources[c.ChainMeta.Id][i].Container.ID)
+					if currentHeight != propHeight {
+						s.T().Logf("current block height on %s is %v, waiting for block %v container: %s", s.valResources[c.ChainMeta.Id][i].Container.Name[1:], currentHeight, propHeight, s.valResources[c.ChainMeta.Id][i].Container.ID)
 					}
-					return syncInfo.SyncInfo.LatestHeight == "75"
+					if currentHeight > propHeight {
+						panic("chain did not halt at upgrade height")
+					}
+					if currentHeight == propHeight {
+						counter++
+					}
+					return counter == 3
```

Comment:
What is the purpose for having both a counter and a 5 minute timeout?

By having a counter running 3 times max every second , we essentially have a 3 second timeout which is much lower than `5*time.Minute`. We can then remove the counter, always return true, and change timeout from `5*time.Minute` to `3*time.Second`

Reply:
The purpose of this was to return True 3 times before accepting it as fact. For instance if upgrade height is 72 its possible to return 72 twice and still hit 73 but 3 seconds is too long to hit the same block 3 times in a row without it having been stopped for upgrade. But this is now moot since we will use a hard chain halt height as suggested by you in your previous comment

Reply:
I think that my earlier proposed solution would functionally the same way as right now, just without a counter. But yeah, if it's going to get change - it doesn't matter :)